### PR TITLE
fix: action cannot be triggered

### PR DESCRIPTION
**Pull request by jnikhila**
- added action cannot be triggered error troubleshooting steps
- fixed linter issues.
As the [PR](https://github.com/appsmithorg/appsmith/pull/17683) for #17683 is still open, the doc PR will also be not merged. But, we can move ahead and do review.

TO-DO --> Add screenshot.
Once the #17683 is available on release, I'll take screenshot and update the PR.

### CODE REVIEW

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (spark-c) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:37]: nit:

`You may see the below errors when executing an ...`

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (spark-c) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:40]: nit:

`You may see this error when ...`

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (spark-c) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:48]: Grammar:

`**Action** refers to the execution of an ...`

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (spark-c) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:51]: I think this :::info might work better when combined with the next sentence ("When you bind an action...") since they are closely related. Maybe by either putting them both in the :::info block, or just having them both be plain text.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (spark-c) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:57]: I think it might help to add a little explanation to clarify why we use `.data`, added in after the first sentence of the **Solution** heading. (It might be hard for new users to understand *why* they need to do this). I put together a little example to show what I mean:


> \#\#\#\# Solution
> You could resolve the error by invoking the `.data` property of the API/query/JS object. Widgets' property fields are synchronous, and they can't be used to call queries (which are async). Instead, once a query has been run, the response is stored as an object on the `<query_name>.data` property. Widgets can accept and work with `<query_name>.data`, since it is just a simple object.
> 
> :::note
> Binding `<query_name>.data` to one of a widget's property fields will make that query run automatically on page load. (This can be switched off in the query's settings page).
> :::
> 
> For example, you have an API ```getFullNameOfLoggedInUser``` that returns a user's...

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md]: @spark-c - As disccused, added the cross-reference for https://docs.appsmith.com/core-concepts/writing-code/workflows#use-case

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md]: True, removed the info.

**Review by spark-c — DISMISSED** [review]

**Review by bharath31 — DISMISSED** [review]

**Comment by jnikhila** [comment]
@bharath31 / @Rishabh-Rathod - Could you please confirm the release date for the [#17683 PR](https://github.com/appsmithorg/appsmith/pull/17683)? So, that we can push the doc changes along.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (Rishabh-Rathod) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: Error messages have been reiterated after a discussion with the team. 

Please check here for more information https://theappsmith.slack.com/archives/C02K0SZQ7V3/p1667457021297869?thread_ts=1667385039.225229&cid=C02K0SZQ7V3

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: @Rishabh-Rathod - This is what I gathered from the conversation. The error message is updated as below:

`Found a Promise() during evaluation. Sync fields cannot execute asynchronous code.`

Please confirm so that I can update the PR.

CC: @bharath31

Comment (Rishabh-Rathod) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: - Below is the error message when an action like storeValue is used in the sync field.

```Found a reference to {{action}} during evaluation. Sync fields cannot execute async framework actions. Please remove any direct/indirect references to {{action}} and try again.```

- When the user writes a JS code in sync fields that returns a promise 

For such cases, as sync fields cannot execute asynchronous code we show the below error message. 

```Found a Promise() during evaluation. Sync fields cannot execute asynchronous code.```

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: Below is the error message when an action like storeValue is used in the sync field.
Error message - 1 --> `Found a reference to {{action}} during evaluation. Sync fields cannot execute async framework actions. Please remove any direct/indirect references to {{action}} and try again.`

Will the `{{action}}` in this example will be replaced with `storeValue`?

For such cases, as sync fields cannot execute asynchronous code we show the below error message.
ErrorMessage - 2 --> `Found a Promise() during evaluation. Sync fields cannot execute asynchronous code.`

A bit confused with the above error messages. For users on the UI which error message will be shown Error message 1 or error message 2.

Comment (Rishabh-Rathod) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: 1. `Will the {{action}} in this example will be replaced with storeValue?`

Yes, to be precise it will be replace with `storeValue()`.  

Example:-
```Found a reference to storeValue() during evaluation. Sync fields cannot execute async framework actions. Please remove any direct/indirect references to storeValue()  and try again.```

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: Thanks @Rishabh-Rathod. And when will the error message 2 be displayed to users.

Comment (Rishabh-Rathod) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: When a user writes code something like ```{{JSObject.myFun1()}}``` in the sync field where `myFun1` is a javascript async function ( function defined with `async` keyword )

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: @Rishabh-Rathod - Do you think we should add these two error messages as two separate troubleshooting errors? CC: @danciaclara - your thoughts??

Comment (Rishabh-Rathod) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:43]: Yes, we should add both errors to help user understand.

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md]: Updated error messages.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:56]: API, Query and JS Objects are cross-referenced once above. So we can skip the cross-referencing again here

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:72]: In this example, let's put a generic name for the API like  `Api1`. Also we don't need to make the user further imagine that this API returns full name of logged in user and all that.  So suggestion to reduce and combine the two lines to 

For example, if you have an API named ```Api1```, you could bind the response generated by the API to a Text widget by referencing the `.data` property as shown in the code snippet below.
```
{{Api1.data}}
```

Also I strongly feel this should be a JSObject example and not with API

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:41]: Just confirming that this error is seen in Query and API also? or just JSObjects @bharath31

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:80]: I feel this line is not required after the example

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:83]: Leave it out of the callout as a normal text

**Review by danciaclara — CHANGES_REQUESTED** [review]

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:5]: The heading are given so that all the errors for that widget can be clubbed under it. For now, JSON form errors just has one error documented 'Source data exceeds`. This would grow as we add more errors.

Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:5]: We can do the restructuring.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:8]: Some error messages are too lengthy to be a part of heading. We can remove Cause and Solution as heading and just have them as bold text if that works.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (jnikhila) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:10]: For all the three. If the Query/API is executed the name of the same will be replaced in {{action}}. However, if it's JS object function then it is Promise() that is shown. CC: @Rishabh-Rathod - Please confirm.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md]: yes, absolutely agree that the JSON Form heading should be there. My point was to club redundant text and reduce unnecessary sub-headings. 
Screenshot 2022-11-26 at 09 48 24

What do you think of the below structure?

## JSON form errors
You may see the below errors when working with a JSON Form widget.

### Source data field limit

**Error Message:** `Source data exceeds 50 fields. Please update the source data` 

This error is seen when you try to bind the query/API response with a lot of fields to the source data property.
Screenshot here

### Cause
...

### Solution
....

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md:8]: The error message heading can be condensed instead of being chopped down eg: **Sync field error** for this one . 
The actual Error message, Cause and Solution should just be bold text as shown in the structure I shared above.
What do you think?

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:25]: nitpick . For some reason, the / is not looking good. 

Wherever this is being done, can we write it with a comma as
API, Query, or JS Object.

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:27]: Maybe this second sentence should start with 'An'

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [inline website/docs/help-and-support/troubleshooting-guide/widget-errors.md:29]: The quotes in the second 'TableData` seems to be incorrect, so it not rendering properly

**Review by danciaclara — CHANGES_REQUESTED** [review]

File: website/docs/help-and-support/troubleshooting-guide/widget-errors.md
Comment (danciaclara) [reply website/docs/help-and-support/troubleshooting-guide/widget-errors.md]: The reason I brought this up is because usually users are aware how to bind the result of query from datasource or API. they know that using .data gives them the result

As for JS Object, when the function is sync, they can directly use {{JSObject.functionName}} in the widget and it works fine. On support queries, I have only ever seen users stuck on this error when the function is async, in which case they have to invoke the function with a .data property.

**Comment by jnikhila** [comment]
@Rishabh-Rathod - I can see that the [PR](https://github.com/appsmithorg/appsmith/pull/17683) for #17683 is closed. However, I cannot see the error message on app.appsmith.com when I try to call async functions in sync fields. Could you please confirm if the ticket has been released?

**Comment by jnikhila** [comment]
@Rishabh-Rathod - Do you think we can publish the doc? Adding this troubleshooting guide shouldn't impact the engineering ticket. CC: @danciaclara 

P.S. - I can do an incremental PR for screenshot once the engineering ticket is released.

**Review by danciaclara — COMMENTED** [review]
As discussed during standup, let's wait for engineering changes to be merged to prod before publishing doc

**Comment by jnikhila** [comment]
@Rishabh-Rathod - Can this PR be merged?

**Comment by Rishabh-Rathod** [comment]
@jnikhila Above PR is blocked by changes in this PR https://github.com/appsmithorg/appsmith/pull/18471. 

I will approve this PR once the blocking PR is merged.

**Comment by Rishabh-Rathod** [comment]
LGTM

**Review by Rishabh-Rathod — APPROVED** [review]

**Pull request merged**
